tutorial/practicals/LFRic/building_code/background/LFRic_PSy.md

```diff
@@ -0,0 +1,223 @@
+
+```fortran
+      INTEGER(KIND=i_def) nlayers
+      ...
+      INTEGER(KIND=i_def), pointer :: map_w0(:,:) => null()
+      INTEGER(KIND=i_def) ndf_w0, undf_w0
+      ...
+      !
+      ! Initialise number of layers
+      !
+      nlayers = field_w0_proxy%vspace%get_nlayers()
+      !
+      ! Look-up dofmaps for each function space
+      !
+      map_w0 => field_w0_proxy%vspace%get_whole_dofmap()
+      !
+      ! Initialise number of DoFs for w0
+      !
+      ndf_w0 = field_w0_proxy%vspace%get_ndf()
+      undf_w0 = field_w0_proxy%vspace%get_undf()
+```
+
+Finally, there is a generated call to `setval_field_w0_code` in the
+`setval_field_w0_kernel_mod` with the complete argument list
+```fortran
+      !
+      ! Call our kernels
+      !
+      DO cell=1,field_w0_proxy%vspace%get_ncell()
+        !
+        CALL setval_field_w0_code(nlayers, field_w0_proxy%data, &
+             scalar_w0, ndf_w0, undf_w0, map_w0(:,cell))
+      END DO
+```
+
+with the same argument order as in the [completed kernel example](
+../1_simple_kernels/solutions/part1/setval_field_w0_kernel_mod.f90)
+provided in the [first example, solutions of Part 1](
+../1_simple_kernels/solutions/part1) (generated by running `genkernelstub`).
+
+The above kernel call code call illustrates how `iterates_over = CELLS`
+metadata in the [`setval_field_w0_kernel_mod.f90`](
+../1_simple_kernels/solutions/part1/setval_field_w0_kernel_mod.f90)
+translates to a **loop over cells** (`cell` loop counter) in the PSy layer.
+
+---
+**NOTE**
+
+LFRic coding standards mandate lower-case for all source including
+Fortran keywords and this holds for all code located in the
+[LFRic repository](LFRic_intro.md#lfric-repository-structure). The
+generated PSyclone Fortran code capitalises keywords for better
+readability (not kept in the LFRic repository).
+
+---
+
+## Generated code to support built-in calls
+
+As illustrated in the [*`invoke` calls* section](
+LFRic_algorithm.md#invoke-calls) of algorithm introduction,
+the `invoke` call from the completed algorithm in the
+[built-ins example solutions](../2_built_ins/solutions/builtins_alg_mod.x90),
+
+```fortran
+    call invoke( name = "Builtins on W0 and W3 fields",              &
+                 ...
+                 X_plus_Y(field_out_w0, field1_in_w0, field2_in_w0), &
+                 ... )
+```
+
+translates to this PSy-layer subroutine call
+
+```fortran
+    CALL invoke_builtins_on_w0_and_w3_fields(field_out_w0, field_out_w3, &
+         field1_in_w0, field2_in_w0, field1_in_w3, field2_in_w3)
+```
+
+in the generated `builtins_alg_mod.f90` file (not kept in the repository).
+
+The generated `invoke_builtins_on_w0_and_w3_fields` subroutine code
+is located in the `builtins_alg_mod_psy.f90` file (also not kept in the
+repository).
+
+As for a kernel call there are, amongst others, generated declarations and
+calls to the `field_out_w0`, `field1_in_w0` and `field2_in_w0` proxies
+
+```fortran
```

Comment:
I will leave it as it is as this part will be going to tutorial 2 on built-ins instead of staying here, so it will be the illustration of what happens with built-in calls in the PSy layer. I added the key point, though.